pkg/util/platformworkloadidentity/platformworkloadidentityrolesbyversion.go

```diff
@@ -0,0 +1,81 @@
+package platformworkloadidentity
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/database"
+	"github.com/Azure/ARO-RP/pkg/util/version"
+)
+
+// PlatformWorkloadIdentityRolesByVersion is the interface that validates and obtains the version from an PlatformWorkloadIdentityRoleSetDocument.
+type PlatformWorkloadIdentityRolesByVersion interface {
+	GetPlatformWorkloadIdentityRolesByRoleName() map[string]api.PlatformWorkloadIdentityRole
+}
+
+// platformWorkloadIdentityRolesByVersionService is the default implementation of the PlatformWorkloadIdentityRolesByVersion interface.
+type platformWorkloadIdentityRolesByVersionService struct {
+	platformWorkloadIdentityRoles []api.PlatformWorkloadIdentityRole
+}
+
+func NewPlatformWorkloadIdentityRolesByVersion(ctx context.Context, oc *api.OpenShiftCluster, dbPlatformWorkloadIdentityRoleSets database.PlatformWorkloadIdentityRoleSets) (PlatformWorkloadIdentityRolesByVersion, error) {
+	if !oc.UsesWorkloadIdentity() {
+		return nil, nil
+	}
+
+	currentInstallVersion := oc.Properties.ClusterProfile.Version
+	currentOpenShiftVersion, err := version.ParseVersion(currentInstallVersion)
+	if err != nil {
+		return nil, err
+	}
+	currentMinorVersion := currentOpenShiftVersion.MinorVersion()
+	requiredMinorVersions := map[string]bool{currentMinorVersion: false}
+	platformWorkloadIdentityRoles := []api.PlatformWorkloadIdentityRole{}
+
+	docs, err := dbPlatformWorkloadIdentityRoleSets.ListAll(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if oc.Properties.PlatformWorkloadIdentityProfile.UpgradeableTo != nil {
+		upgradeableVersion, err := version.ParseVersion(string(*oc.Properties.PlatformWorkloadIdentityProfile.UpgradeableTo))
+		if err != nil {
+			return nil, err
+		}
+		upgradeableMinorVersion := upgradeableVersion.MinorVersion()
+		if currentMinorVersion != upgradeableMinorVersion && !currentOpenShiftVersion.Lt(upgradeableVersion) {
```

Comment:
Do we want to handle cases where `currentOpenShiftVersion.Lt(upgradeableVersion)` or just ignore it?

Reply:
We can ignore it, as it will mean that the cluster is upgraded and will be ignored as part of the validation.

Reply:
We might have an opportunity for reusable logic around "what OCP version do I need to handle upgradeability to" across multiple steps in our cluster update process. I think we might benefit from extracting this logic here into a common function and setting up these kinds of checks so we have a single consistent version:
- UpgradeableTo not specified (return the current version)
- UpgradeableTo specified but less than current version (return the current version)
- UpgradeableTo specified and greater than current version (return the UpgradeableTo value)